pkg/frontend/querymiddleware/codec.go

```diff
@@ -112,7 +116,45 @@ type Response interface {
 	GetHeaders() []*PrometheusResponseHeader
 }
 
-type prometheusCodec struct{}
+type prometheusCodecMetrics struct {
+	duration *prometheus.HistogramVec
+	size     *prometheus.HistogramVec
+}
+
+func newPrometheusCodecMetrics(registerer prometheus.Registerer) *prometheusCodecMetrics {
+	factory := promauto.With(registerer)
+	ms := 1.0 / 1000
+	us := ms / 1000
+	kb := 1024.0
+	mb := 1024 * kb
+
+	return &prometheusCodecMetrics{
+		duration: factory.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: "cortex",
+			Name:      "frontend_query_payload_codec_duration_seconds",
+			Help:      "Total time spent encoding or decoding query result payloads, in seconds.",
+			// We need a wide range of buckets here because encoding protobuf payloads can take as little as 80ns, and decoding very large JSON payloads can take many milliseconds.
+			Buckets: []float64{1 * us, 10 * us, 100 * us, 500 * us, 1 * ms, 5 * ms, 10 * ms, 20 * ms},
```

Comment:
Do we really care about anything below 1ms? I also think you've been too much optimistic on the encoding/decoding time. Under some edge cases we see hundred MBs query response, which may take way longer than 20ms.

Reply:
Resolved in 6c41f95. We now have a set of 10 exponentially increasing buckets from 1 ms to 2 seconds.